// react-aria/lib/Accordion/Tab.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _specialAssign = require('../special-assign');

var _specialAssign2 = _interopRequireDefault(_specialAssign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var checkedProps = {
  tag: _react.PropTypes.string,
  id: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.number]).isRequired,
  isActive: _react.PropTypes.bool,
  children: _react.PropTypes.oneOfType([_react.PropTypes.func, _react.PropTypes.node]).isRequired
};

var Tab = function (_Component) {
  _inherits(Tab, _Component);

  function Tab() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Tab);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Tab.__proto__ || Object.getPrototypeOf(Tab)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      isActive: false
    }, _this._setActiveState = function (isActive) {
      _this.setState({ isActive: isActive });
    }, _this._toggleActiveState = function () {
      _this.setState({ isActive: !_this.state.isActive });
    }, _this._handleKeyDown = function (e) {
      var _this$context$ariaMan = _this.context.ariaManager;
      var type = _this$context$ariaMan.type;
      var activateTab = _this$context$ariaMan.activateTab;
      var _this$props = _this.props;
      var id = _this$props.id;
      var onKeyDown = _this$props.onKeyDown;


      if (type === 'accordion' && [' ', 'Enter'].indexOf(e.key) > -1) {
        activateTab(id);
      }
      if (onKeyDown) {
        onKeyDown(e);
      }
    }, _this._handleFocus = function (e) {
      var _this$context$ariaMan2 = _this.context.ariaManager;
      var type = _this$context$ariaMan2.type;
      var activateTab = _this$context$ariaMan2.activateTab;
      var _this$props2 = _this.props;
      var id = _this$props2.id;
      var onFocus = _this$props2.onFocus;


      if (type === 'tabs') {
        activateTab(id);
      }
      if (onFocus) {
        onFocus(e);
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Tab, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this._member = {
        type: 'tab',
        id: this.props.id,
        node: (0, _reactDom.findDOMNode)(this),
        text: this.props.text,
        setActiveState: this._setActiveState,
        toggleActiveState: this._toggleActiveState
      };
      this.context.ariaManager.addMember(this._member);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.context.ariaManager.removeMember(this._member);
    }
  }, {
    key: 'render',
    value: function render() {
      var type = this.context.ariaManager.type;
      var _props = this.props;
      var tag = _props.tag;
      var id = _props.id;
      var disabled = _props.disabled;
      var children = _props.children;

      var isActive = this.props.isActive !== undefined ? this.props.isActive : this.state.isActive;
      var componentProps = {
        id: id,
        role: 'tab',
        tabIndex: type === 'accordion' ? 0 : isActive ? 0 : -1,
        'aria-selected': isActive,
        'aria-controls': id + '-panel',
        'aria-disabled': disabled,
        onKeyDown: this._handleKeyDown,
        onFocus: this._handleFocus
      };

      if (type === 'accordion') {
        componentProps['aria-expanded'] = isActive;
      }

      var props = (0, _specialAssign2.default)(componentProps, this.props, checkedProps);

      if (typeof children === 'function') {
        return children(props, isActive);
      }

      return (0, _react.createElement)(tag, props, children);
    }
  }]);

  return Tab;
}(_react.Component);

Tab.contextTypes = {
  ariaManager: _react.PropTypes.object.isRequired
};
Tab.propTypes = checkedProps;
Tab.defaultProps = {
  tag: 'div'
};
exports.default = Tab;
module.exports = exports['default'];
