// react-aria/lib/helpers/Members.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _FocusGroup = require('./FocusGroup');

var _FocusGroup2 = _interopRequireDefault(_FocusGroup);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Members = function Members() {
  var _this = this;

  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  _classCallCheck(this, Members);

  this.add = function (member) {
    var index = member.index;
    var node = member.node;
    var text = member.text;


    if (index === undefined) {
      _this.collection.push(member);
    } else {
      _this.collection.splice(index, 0, member);
    }

    // add member to focus group
    _this.focusGroup.addMember({
      node: node,
      text: text || node.innerHTML
    });

    // activate focus group if this was the first member added
    if (_this.collection.length === 1) {
      _this.focusGroup.activate();
    }

    if (typeof _this.options.onAdd === 'function') {
      _this.options.onAdd(member);
    }
  };

  this.remove = function (member) {
    var pos = _this.collection.indexOf(member);

    if (pos > -1) {
      _this.collection.splice(member, 1);
      _this.focusGroup.removeMember(member.node);
    }

    // deactivate focus group if this was the last member removed
    if (_this.collection.length <= 0) {
      _this.focusGroup.deactivate();
    }

    if (typeof _this.options.onRemove === 'function') {
      _this.options.onRemove(member);
    }
  };

  this.focus = function (index) {
    _this.focusGroup.focus(index);
  };

  this.collection = [];
  this.options = options;
  this.focusGroup = new _FocusGroup2.default(this.options);
};

exports.default = Members;
module.exports = exports['default'];
