// react-aria/lib/ComboBox/ListBox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _Overlays = require('../Overlays');

var _specialAssign = require('../utils/special-assign');

var _specialAssign2 = _interopRequireDefault(_specialAssign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var checkedProps = {
  tag: _react.PropTypes.string,
  children: _react.PropTypes.node,
  closeOnOutsideClick: _react.PropTypes.bool,
  onItemHighlight: _react.PropTypes.func
};

var ListBox = function (_Component) {
  _inherits(ListBox, _Component);

  function ListBox() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ListBox);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ListBox.__proto__ || Object.getPrototypeOf(ListBox)).call.apply(_ref, [this].concat(args))), _this), _this._handleItemHightlight = function (item) {
      _this.context.comboBox.setActiveDescendant(item);

      if (typeof _this.props.onItemHighlight === 'function') {
        _this.props.onItemHighlight(item);
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ListBox, [{
    key: 'render',
    value: function render() {
      var comboBox = this.context.comboBox;
      var _props = this.props;
      var tag = _props.tag;
      var closeOnOutsideClick = _props.closeOnOutsideClick;
      var children = _props.children;

      var props = (0, _specialAssign2.default)({
        role: 'listbox',
        id: comboBox.uuid,
        rootNode: comboBox.inputNode,
        closeOnOutsideClick: closeOnOutsideClick,
        onItemHighlight: this._handleItemHightlight
      }, this.props, checkedProps);

      return (0, _react.createElement)(_Overlays.Overlay, props, children);
    }
  }]);

  return ListBox;
}(_react.Component);

ListBox.contextTypes = {
  comboBox: _react.PropTypes.object
};
ListBox.propTypes = checkedProps;
ListBox.defaultProps = {
  tag: 'div',
  closeOnOutsideClick: false
};
exports.default = ListBox;
module.exports = exports['default'];
