// react-aria/lib/Overlays/Input.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _specialAssign = require('../helpers/special-assign');

var _specialAssign2 = _interopRequireDefault(_specialAssign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var checkedProps = {
  overlayRole: _react.PropTypes.string,
  controls: _react.PropTypes.string,
  isOpen: _react.PropTypes.bool,
  children: _react.PropTypes.func
};

var Trigger = function (_Component) {
  _inherits(Trigger, _Component);

  function Trigger() {
    _classCallCheck(this, Trigger);

    return _possibleConstructorReturn(this, (Trigger.__proto__ || Object.getPrototypeOf(Trigger)).apply(this, arguments));
  }

  _createClass(Trigger, [{
    key: '_getProps',
    value: function _getProps() {
      var _props = this.props;
      var overlayRole = _props.overlayRole;
      var controls = _props.controls;
      var isOpen = _props.isOpen;

      var props = {};

      if (overlayRole !== 'modal') {
        props['aria-haspopup'] = true;
        props['aria-expanded'] = isOpen;
      }

      if (overlayRole === 'popover') {
        props['id'] = controls;
      } else if (overlayRole === 'tooltip') {
        props['aria-describedby'] = controls;
      }

      return (0, _specialAssign2.default)(props, this.props, checkedProps);
    }
  }, {
    key: 'render',
    value: function render() {
      var children = this.props.children;

      var props = this._getProps();

      if (typeof children === 'function') {
        return children(props);
      }

      return (0, _react.createElement)('input', props);
    }
  }]);

  return Trigger;
}(_react.Component);

Trigger.propTypes = checkedProps;
Trigger.defaultProps = {
  overlayRole: 'popover'
};
exports.default = Trigger;
module.exports = exports['default'];
