// react-aria/lib/Overlays/ItemCollection.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _a11yFocusScope = require('a11y-focus-scope');

var _Members = require('../utils/Members');

var _Members2 = _interopRequireDefault(_Members);

var _specialAssign = require('../utils/special-assign');

var _specialAssign2 = _interopRequireDefault(_specialAssign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var checkedProps = {
  tag: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.bool]),
  rootNode: _react.PropTypes.any,
  scopeFocus: _react.PropTypes.bool,
  initialFocus: _react.PropTypes.any,
  onItemFocus: _react.PropTypes.func,
  onItemSelection: _react.PropTypes.func,
  children: _react.PropTypes.oneOfType([_react.PropTypes.func, _react.PropTypes.node])
};
var noop = function noop() {
  return null;
};

var ItemCollection = function (_Component) {
  _inherits(ItemCollection, _Component);

  function ItemCollection() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ItemCollection);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ItemCollection.__proto__ || Object.getPrototypeOf(ItemCollection)).call.apply(_ref, [this].concat(args))), _this), _this._members = new _Members2.default(), _this.getItems = function () {
      return _this._members;
    }, _this.focusItem = function (index) {
      _this._members.focus(index);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ItemCollection, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        itemCollection: {
          members: this._members,
          onItemSelection: this.props.onItemSelection
        }
      };
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this._lastActiveElement = document.activeElement;

      this._members.on('focus', this.props.onItemFocus);
      this._members.on('select', this.props.onItemSelection);

      if (this.props.rootNode) {
        this._members.setRootNode(this.props.rootNode);
      }

      if (this.props.scopeFocus) {
        (0, _a11yFocusScope.scopeFocus)((0, _reactDom.findDOMNode)(this));
      }

      if (this.props.initialFocus) {
        this._members.focus(0);
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (this.props.scopeFocus) {
        (0, _a11yFocusScope.unscopeFocus)();
      }

      if (this._lastActiveElement) {
        this._lastActiveElement.focus();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props;
      var tag = _props.tag;
      var children = _props.children;

      var props = (0, _specialAssign2.default)({}, this.props, checkedProps);

      if (!tag) {
        return children;
      }

      return (0, _react.createElement)(tag, props, children);
    }
  }]);

  return ItemCollection;
}(_react.Component);

ItemCollection.childContextTypes = {
  itemCollection: _react.PropTypes.object
};
ItemCollection.propTypes = checkedProps;
ItemCollection.defaultProps = {
  tag: false,
  initialFocus: true,
  onItemFocus: noop,
  onItemSelection: noop
};
exports.default = ItemCollection;
module.exports = exports['default'];
