// react-aria/lib/Overlays/state-manager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registerPopover = registerPopover;
exports.unregisterPopover = unregisterPopover;
exports.getPopover = getPopover;
exports.openPopover = openPopover;
exports.closePopover = closePopover;
exports.togglePopover = togglePopover;
exports.registerToggle = registerToggle;
exports.unregisterToggle = unregisterToggle;
exports.getToggle = getToggle;
var toggles = {};
var popovers = {};
var members = {};

function registerPopover(id, popover) {
  popovers[id] = popover;
}

function unregisterPopover(id, popover) {
  delete popovers[id];
}

function getPopover(id) {
  return popovers[id];
}

function openPopover(id, options) {
  popovers[id].open(options);
}

function closePopover(id, options) {
  popovers[id].close(options);
}

function togglePopover(id, options) {
  popovers[id].toggle(options);
}

function registerToggle(id, toggle) {
  toggles[id] = toggle;
}

function unregisterToggle(id) {
  delete toggles[id];
}

function getToggle(id) {
  return toggles[id];
}
