// react-aria/lib/AriaToggle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _specialAssign = require('./special-assign');

var _specialAssign2 = _interopRequireDefault(_specialAssign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var checkedProps = {
  tag: _react.PropTypes.string,
  children: _react.PropTypes.oneOfType([_react.PropTypes.func, _react.PropTypes.node]).isRequired
};

var AriaToggle = function (_Component) {
  _inherits(AriaToggle, _Component);

  function AriaToggle() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, AriaToggle);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = AriaToggle.__proto__ || Object.getPrototypeOf(AriaToggle)).call.apply(_ref, [this].concat(args))), _this), _this._handleKeyDown = function (e) {
      if (['ArrowUp', 'ArrowDown', ' '].indexOf(e.key) > -1 || _this.props.tag !== 'button' && e.key === 'Enter') {
        if (!_this.context.ariaManager.isPopoverOpen) {
          _this.context.ariaManager.openPopover();
        } else {
          _this.context.ariaManager.focusItem(0);
        }
      }
      if (_this.props.onKeyDown) {
        _this.props.onKeyDown(e);
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(AriaToggle, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.context.ariaManager.setToggleNode((0, _reactDom.findDOMNode)(this));
    }
  }, {
    key: 'render',
    value: function render() {
      var _context$ariaManager = this.context.ariaManager;
      var type = _context$ariaManager.type;
      var uuid = _context$ariaManager.uuid;
      var isPopoverOpen = _context$ariaManager.isPopoverOpen;
      var _props = this.props;
      var tag = _props.tag;
      var disabled = _props.disabled;
      var children = _props.children;

      var componentProps = {
        role: 'button',
        tabIndex: disabled ? '' : 0,
        'aria-haspopup': true,
        'aria-expanded': isPopoverOpen,
        'aria-disabled': disabled,
        onKeyDown: this._handleKeyDown
      };

      if (type === 'popover') {
        componentProps['id'] = uuid;
      }

      if (type === 'tooltip') {
        componentProps['aria-describedby'] = uuid;
      }

      var props = (0, _specialAssign2.default)(componentProps, this.props, checkedProps);

      if (typeof children === 'function') {
        return children(props);
      }

      return (0, _react.createElement)(tag, props, children);
    }
  }]);

  return AriaToggle;
}(_react.Component);

AriaToggle.contextTypes = {
  ariaManager: _react.PropTypes.object.isRequired
};
AriaToggle.propTypes = checkedProps;
AriaToggle.defaultProps = {
  tag: 'button'
};
exports.default = AriaToggle;
module.exports = exports['default'];
