// react-aria/lib/AriaTabList.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _specialAssign = require('./special-assign');

var _specialAssign2 = _interopRequireDefault(_specialAssign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var checkedProps = {
  tag: _react.PropTypes.string,
  children: _react.PropTypes.oneOfType([_react.PropTypes.func, _react.PropTypes.node]).isRequired
};

var AriaTabList = function (_Component) {
  _inherits(AriaTabList, _Component);

  function AriaTabList() {
    _classCallCheck(this, AriaTabList);

    return _possibleConstructorReturn(this, (AriaTabList.__proto__ || Object.getPrototypeOf(AriaTabList)).apply(this, arguments));
  }

  _createClass(AriaTabList, [{
    key: 'render',
    value: function render() {
      var type = this.context.ariaManager.type;
      var _props = this.props;
      var tag = _props.tag;
      var children = _props.children;

      var componentProps = {
        role: 'tablist'
      };

      if (type === 'accordion') {
        componentProps['aria-multiselectable'] = true;
      }

      var props = (0, _specialAssign2.default)(componentProps, this.props, checkedProps);

      if (typeof children === 'function') {
        return children(props);
      }

      return (0, _react.createElement)(tag, props, children);
    }
  }]);

  return AriaTabList;
}(_react.Component);

AriaTabList.contextTypes = {
  ariaManager: _react.PropTypes.object.isRequired
};
AriaTabList.propTypes = checkedProps;
AriaTabList.defaultProps = {
  tag: 'div'
};
exports.default = AriaTabList;
module.exports = exports['default'];
