// react-aria/lib/AriaPanel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _specialAssign = require('./special-assign');

var _specialAssign2 = _interopRequireDefault(_specialAssign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var checkedProps = {
  tag: _react.PropTypes.string,
  controlledBy: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.number]).isRequired,
  isActive: _react.PropTypes.bool,
  children: _react.PropTypes.oneOfType([_react.PropTypes.func, _react.PropTypes.node]).isRequired
};

var AriaPanel = function (_Component) {
  _inherits(AriaPanel, _Component);

  function AriaPanel() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, AriaPanel);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = AriaPanel.__proto__ || Object.getPrototypeOf(AriaPanel)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      isActive: false
    }, _this._setActiveState = function (isActive) {
      _this.setState({ isActive: isActive });
    }, _this._toggleActiveState = function () {
      _this.setState({ isActive: !_this.state.isActive });
    }, _this._handleKeyDown = function (e) {
      if (e.ctrlKey && e.key === 'ArrowUp') {
        _this.context.ariaManager.focusTab(_this.props.controlledBy);
      }
      if (_this.props.onKeyDown) {
        _this.props.onKeyDown(e);
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(AriaPanel, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.context.ariaManager.addPanel({
        controlledBy: this.props.controlledBy,
        node: (0, _reactDom.findDOMNode)(this),
        setActiveState: this._setActiveState,
        toggleActiveState: this._toggleActiveState
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props;
      var tag = _props.tag;
      var controlledBy = _props.controlledBy;
      var disabled = _props.disabled;
      var children = _props.children;

      var isActive = this.props.isActive !== undefined ? this.props.isActive : this.state.isActive;
      var componentProps = {
        id: controlledBy + '-panel',
        role: 'tabpanel',
        'aria-hidden': !isActive,
        'aria-labelledby': controlledBy,
        onKeyDown: this._handleKeyDown
      };

      if (!isActive) {
        componentProps['style'] = _extends({
          display: 'none'
        }, this.props.style);
      }

      var props = (0, _specialAssign2.default)(componentProps, this.props, checkedProps);

      if (typeof children === 'function') {
        return children(props, isActive);
      }

      return (0, _react.createElement)(tag, props, children);
    }
  }]);

  return AriaPanel;
}(_react.Component);

AriaPanel.contextTypes = {
  ariaManager: _react.PropTypes.object.isRequired
};
AriaPanel.propTypes = checkedProps;
AriaPanel.defaultProps = {
  tag: 'div'
};
exports.default = AriaPanel;
module.exports = exports['default'];
